py/selenium/webdriver/firefox/webdriver.py

```diff
@@ -47,35 +47,99 @@ class WebDriver(RemoteWebDriver):
 
     _web_element_cls = FirefoxWebElement
 
-    def __init__(self, firefox_profile=None, firefox_binary=None, timeout=30,
-                 capabilities=None, proxy=None, executable_path="geckodriver", firefox_options=None, log_path="geckodriver.log"):
-        capabilities = capabilities or DesiredCapabilities.FIREFOX.copy()
-
-        self.profile = firefox_profile or FirefoxProfile()
-        self.profile.native_events_enabled = (
-            self.NATIVE_EVENTS_ALLOWED and self.profile.native_events_enabled)
-
-        self.binary = firefox_binary or capabilities.get("binary", FirefoxBinary())
-
-        self.options = firefox_options or Options()
-        self.options.binary_location = self.binary if isinstance(self.binary, basestring) else self.binary._start_cmd
-        self.options.profile = self.profile
-        capabilities.update(self.options.to_capabilities())
-
-        # marionette
+    def __init__(self, firefox_profile=None, firefox_binary=None,
+                 timeout=30, capabilities=None, proxy=None,
+                 executable_path="geckodriver", firefox_options=None,
+                 log_path="geckodriver.log"):
+        """Starts a new local session of Firefox.
+
+        Based on the combination and specificity of the various keyword
+        arguments, a capabilities dictionary will be constructed that
+        is passed to the remote end.
+
+        The keyword arguments given to this constructor are helpers to
+        more easily allow Firefox WebDriver sessions to be customised
+        with different options.  They are mapped on to a capabilities
+        dictionary that is passed on to the remote end.
+
+        As some of the options, such as `firefox_profile` and
+        `firefox_options.profile` are mutually exclusive, precedence is
+        given from how specific the setting is.  `capabilities` is the
+        least specific keyword argument, followed by `firefox_options`,
+        followed by `firefox_binary` and `firefox_profile`.
+
+        In practice this means that if `firefox_profile` and
+        `firefox_options.profile` are both set, the selected profile
+        instance will always come from the most specific variable.
+        In this case that would be `firefox_profile`.  This will result in
+        `firefox_options.profile` to be ignored because it is considered
+        a less specific setting than the top-level `firefox_profile`
+        keyword argument.  Similarily, if you had specified a
+        `capabilities["firefoxOptions"]["profile"]` Base64 string,
+        this would rank below `firefox_options.profile`.
+
+        :param firefox_profile: Instance of ``FirefoxProfile`` object.
+        :param firefox_binary: Full path to the Firefox binary to use.
```

Comment:
The current implementation supports either a string or a `FirefoxBinary` instance.